### PR TITLE
Snapshotting pseudo state & log entry cache

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request contains multiple changes/fixes for the `raft` package:

* Manage requests better while snapshotting. Currently a large snapshot will lock the log and queue up heartbeats & possibly cause an election timeout when it's done. There is now a `Log.snapshotting` field that is set using the `sync/atomic` package.

* Add a log entry cache so that temporarily disconnected streams can reconnect without requiring a snapshot.

* Wrap the `Log.mu` mutex with `lock()` and `unlock()` methods for lock contention debugging. There's also a `printCaller()` method that will log the file & line of the calling function to a lock.

### CODE REVIEW

File: raft/log.go [hunk 1]
```diff
@@ -1780,6 +1829,8 @@ func (l *Log) ReadFrom(r io.ReadCloser) error {
 		return nil
 	}
 
+	l.Logger.Printf("reading from stream")
```
Comment (corylanou) [inline raft/log.go]: Did you still want this for debugging?

Comment (corylanou) [inline raft/log.go]: nm, I see a bunch of them below, assuming this is needed for debugging in the wild.

File: raft/log.go [hunk 2]
```diff
@@ -203,38 +218,48 @@ func NewLog() *Log {
...
+
+func (l *Log) printCaller(label string) {
+	_, file, line, _ := runtime.Caller(2)
+	l.Logger.Printf("%s: %s:%d", label, filepath.Base(file), line)
+}
+
 // Path returns the data path of the Raft log.
 // Returns an empty string if the log is closed.
 func (l *Log) Path() string {
-	l.mu.Lock()
```
Comment (otoolep) [inline raft/log.go]: Why this cosmetic change?

Comment (otoolep) [inline raft/log.go]: OK, you called it out in your summary -- thanks.

File: raft/log.go [hunk 3]
```diff
@@ -867,18 +884,26 @@ func (l *Log) followerLoop(closing <-chan struct{}) State {
 			return Stopped
 		case ch := <-l.Clock.AfterElectionTimeout():
 			close(ch)
+
+			// Ignore timeout if we are snapshotting.
```
Comment (otoolep) [inline raft/log.go]: Timeout will be set, and tried again, right?

File: raft/log.go [hunk 4]
```diff
@@ -1578,8 +1614,14 @@ func (l *Log) Heartbeat(term, commitIndex, leaderID uint64) (currentIndex uint64
 
 // RequestVote requests a vote from the log.
 func (l *Log) RequestVote(term, candidateID, lastLogIndex, lastLogTerm uint64) (peerTerm uint64, err error) {
-	l.mu.Lock()
-	defer l.mu.Unlock()
+	// Ignore if snapshotting.
```
Comment (otoolep) [inline raft/log.go]: What will this do to leader election? It may cause leader election to be restarted, right, if not enough votes were received?

**Comment by otoolep** [comment]
Interesting change, generally makes sense. I do have questions about how the system will respond if stuff that used to happen anyway, is now explicitly blocked by an active snapshot.

File: raft/log.go (hunk 4 above)
Comment (benbjohnson) [inline raft/log.go]: Yes, it'll act the same as if the node was down since it sort of is.

**Comment by otoolep** [comment]
+1, change generally makes sense.

**Comment by toddboom** [comment]
:+1: from me too

**Pull request merged**